src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs

```diff
@@ -802,38 +802,59 @@ internal bool Synchronize(ChainedHeader newTip)
+                        if (this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
                         {
-                            if (this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
-                            {
-                                this.logger.Trace("(-)[NODE_DISPOSED]");
-                                this.PollsRepository.SaveCurrentTip(transaction);
-                                transaction.Commit();
+                            this.logger.Trace("(-)[NODE_DISPOSED]");
+                            this.PollsRepository.SaveCurrentTip(currentTransaction);
+                            currentTransaction.Commit();
+                            currentTransaction.Dispose();
 
-                                bSuccess = false;
-                                return;
-                            }
+                            bSuccess = false;
+                            return;
+                        }
 
-                            ChainedHeader header = headers[i++];
-                            this.ProcessBlock(transaction, new ChainedHeaderBlock(block, header));
+                        ChainedHeader header = headers[i++];
 
-                            if (header.Height % 10000 == 0)
-                            {
-                                var progress = (int)((decimal)header.Height / this.chainIndexer.Tip.Height * 100);
-                                var progressString = $"Synchronizing voting data at height {header.Height} / {this.chainIndexer.Tip.Height} ({progress} %).";
+                        this.ProcessBlock(currentTransaction, new ChainedHeaderBlock(block, header));
 
-                                this.logger.Info(progressString);
-                                this.signals.Publish(new FullNodeEvent() { Message = progressString, State = FullNodeState.Initializing.ToString() });
-                            }
+                        if (header.Height % 10000 == 0)
+                        {
+                            // Start a new transaction.
+                            if (currentTransactionCommitted)
+                                currentTransaction = this.PollsRepository.GetTransaction();
+
+                            var progress = (int)((decimal)header.Height / this.chainIndexer.Tip.Height * 100);
+                            var progressString = $"Synchronizing voting data at height {header.Height} / {this.chainIndexer.Tip.Height} ({progress} %).";
+
+                            this.logger.Info(progressString);
+                            this.signals.Publish(new FullNodeEvent() { Message = progressString, State = FullNodeState.Initializing.ToString() });
+
+                            this.PollsRepository.SaveCurrentTip(currentTransaction);
+
+                            currentTransaction.Commit();
+                            currentTransaction.Dispose();
+
+                            currentTransactionCommitted = true;
```

Comment:
Maybe just re-open the transaction instead of having this variable

Reply:
Done

Reply:
I did initially do this but then realized I had no way of knowing whether or not the transaction was committed when you exit the loop. `Dbreeze.Transaction` does not have a `IsCommitted` property or something similar.

Reply:
If you always reopen the transaction then you can always assume its not committed